datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -1046,6 +1059,130 @@ public int cleanUpOrphanConnectorTasks(boolean cleanUpOrphanTasksInConnector) {
+    // do not perform this operation if not a leader or another operation is going on.
+    if (!_isLeader || !_orphanLockCleanupFuture.isDone()) {
+      return 0;
+    }
+
+    Set<DatastreamTask> validTasks = getDatastreamTasks();
+    List<String> allConnectors = getAllConnectors();
+
+    int orphanCount = 0;
+    for (String connector : allConnectors) {
+      Map<String, Set<String>> locksByTaskPrefix = getAllConnectorTaskLocks(connector);
+
+      Set<String> validTaskNamesSet = validTasks.stream()
+          .filter(x -> x.getConnectorType().equals(connector))
+          .map(DatastreamTask::getDatastreamTaskName)
+          .collect(Collectors.toSet());
+
+      List<String> orphanLockList = new ArrayList<>();
+
+      locksByTaskPrefix.forEach((taskPrefix, taskList) -> {
+        // if all the task locks in the task prefix needs to be cleaned up, delete the task prefix node as well.
+        AtomicBoolean deleteTaskPrefixNode = new AtomicBoolean(true);
+        taskList.forEach(taskName -> {
+          if (!validTaskNamesSet.contains(taskName)) {
+            orphanLockList.add(KeyBuilder.datastreamTaskLock(_cluster, connector, taskPrefix, taskName));
+          } else {
+            deleteTaskPrefixNode.set(false);
+          }
+        });
+
+        if (deleteTaskPrefixNode.get()) {
+          orphanLockList.add(KeyBuilder.datastreamTaskLockPrefix(_cluster, connector, taskPrefix));
+        }
+      });
+
+      if (orphanLockList.size() > 0) {
+        LOG.warn("Found orphan task locks: {} in connector: {}", orphanLockList, connector);
+        if (cleanUpOrphanTaskLocksInConnector) {
+          _finalOrphanLockList.addAll(orphanLockList);
+        }
+        orphanCount += orphanLockList.size();
+      }
+    }
+
+    if (cleanUpOrphanTaskLocksInConnector) {
+      // waiting for the debounce time to ensure that the task thread should stop processing by then.
+      _orphanLockCleanupFuture = _scheduledExecutorServiceOrphanLockCleanup.schedule(this::cleanUpOrphanLocks,
+          _debounceTimerMs, TimeUnit.MILLISECONDS);
+    }
+    return orphanCount;
+  }
+
+  private void cleanUpOrphanLocks() {
+    _finalOrphanLockList.forEach(t -> {
+      LOG.info("Deleting task lock node {}", t);
+      _zkclient.delete(t);
+    });
+    _finalOrphanLockList.clear();
+  }
+
+  /*
+   * wait for the task lock to release and delete if the lock is owned by dead owner
+   */
+  private void waitForTaskReleaseOrForceIfOwnerIsDead(DatastreamTask task, long timeoutMs, String lockPath) {
+    boolean deadOwner = false;
+    if (_zkclient.exists(lockPath)) {
+      String owner = _zkclient.readData(lockPath, true);
+      if (owner != null && owner.equals(_instanceName)) {
+        LOG.info("{} already owns the lock on {}", _instanceName, task);
+        //release the lock.
+        return;
+      }
+
+      long waitTimeout = timeoutMs;
+      // if the timeout is greater than debounce timer, only then identify and clean the task lock.
```

Comment:
I also have the same feeling that should we make this timeout requirement greater than debounceTimer. Keeping this open for discussion, but wanted to add the support in case we decide the other way to not tie debounceTimer and acquire timer.

Reply:
Sure, let's see what @ahmedahamid thinks about this API and make a call either way.